ext/lists.go

```diff
@@ -276,29 +450,122 @@ func flatten(list traits.Lister, depth int64) ([]ref.Val, error) {
 	if _, ok := elem.(traits.Comparer); !ok {
 		return nil, fmt.Errorf("list elements must be comparable")
 	}
 
+	sortedIndices := make([]ref.Val, 0, listLength)
+	for i := types.IntZero; i < listLength; i++ {
+		if keys.Get(i).Type() != elem.Type() {
+			return nil, fmt.Errorf("list elements must have the same type")
+		}
+		sortedIndices = append(sortedIndices, i)
+	}
+
+	sort.Slice(sortedIndices, func(i, j int) bool {
+		iKey := keys.Get(sortedIndices[i])
+		jKey := keys.Get(sortedIndices[j])
+		return iKey.(traits.Comparer).Compare(jKey) == types.IntNegOne
+	})
+
 	sorted := make([]ref.Val, 0, listLength)
+	for i := types.IntZero; i < listLength; i++ {
+		sorted = append(sorted, list.Get(sortedIndices[i]))
+	}
+	return types.DefaultTypeAdapter.NativeToValue(sorted), nil
+}
+
+// sortByMacro transforms an expression like:
+//
+//	mylistExpr.sortBy(e, -math.abs(e))
+//
+// into something equivalent to:
+//
+//	 cel.bind(
+//		  __sortBy_input__,
+//		  myListExpr,
+//		  __sortBy_input__.sortByAssociatedKeys(__sortBy_input__.map(e, -math.abs(e))
+//		)
+func sortByMacro(meh cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	varIdent := meh.NewIdent("__sortBy_input__")
+	varName := varIdent.AsIdent()
+
+	targetKind := target.Kind()
+	if targetKind != ast.ListKind &&
+		targetKind != ast.SelectKind &&
+		targetKind != ast.IdentKind &&
+		targetKind != ast.ComprehensionKind && targetKind != ast.CallKind {
+		return nil, meh.NewError(target.ID(), fmt.Sprintf("sortBy can only be applied to a list, identifier, comprehension, call or select expression"))
+	}
+
+	mapCompr, err := parser.MakeMap(meh, varIdent, args)
+	if err != nil {
+		return nil, err
+	}
+	callExpr := meh.NewMemberCall("sortByAssociatedKeys",
+		varIdent,
+		mapCompr,
+	)
+
+	bindExpr := meh.NewComprehension(
+		meh.NewList(),
+		"#unused",
+		varName,
+		target,
+		meh.NewLiteral(types.False),
+		varIdent,
+		callExpr,
+	)
+
+	return bindExpr, nil
+}
+
+func distinctList(list traits.Lister) (ref.Val, error) {
+	listLength := list.Size().(types.Int)
+	if listLength == 0 {
+		return list, nil
+	}
+	elem := list.Get(types.IntZero)
+	if _, ok := elem.(traits.Comparer); !ok {
+		return nil, fmt.Errorf("list elements must be comparable")
+	}
+
+	uniqueList := make([]ref.Val, 0, listLength)
+	seen := make(map[ref.Val]bool, listLength)
 	for i := types.IntZero; i < listLength; i++ {
 		val := list.Get(i)
 		if val.Type() != elem.Type() {
 			return nil, fmt.Errorf("list elements must have the same type")
 		}
-		sorted = append(sorted, val)
+		if !seen[val] {
+			uniqueList = append(uniqueList, val)
+			seen[val] = true
+		}
 	}
 
-	sort.Slice(sorted, func(i, j int) bool {
-		return sorted[i].(traits.Comparer).Compare(sorted[j]) == types.IntNegOne
+	// Sort for determinism across implementations.
+	sort.Slice(uniqueList, func(i, j int) bool {
+		return uniqueList[i].(traits.Comparer).Compare(uniqueList[j]) == types.IntNegOne
 	})
```

Comment:
I would prefer if we left the distinct check as unsorted by default, but rather treated as an O(n^2) insertion based on equality. In SQL, a `SELECT DISTINCT` does not have ordering implications, but rather preserves the record order as it was presented.